src/plugins/discord/mirror.js

```diff
@@ -0,0 +1,384 @@
+// @flow
+
+import type Database from "better-sqlite3";
+import stringify from "json-stable-stringify";
+import * as Model from "./models";
+import dedent from "../../util/dedent";
+
+const VERSION = "DISCORD_MIRROR_v1";
+
+/**
+ * Persists a local copy of data from a Discord Guild.
+ * Implements create and read functionality.
+ *
+ * Each Mirror instance is tied to a particular Guild. Trying to use a mirror
+ * for multiple Discord Guilds is not permitted; use separate Mirrors.
+ */
+
+export class Mirror {
+  +_db: Database;
+
+  /**
+   * Construct a new Mirror instance.
+   *
+   * Takes a Database and GuildId.
+   *
+   */
+
+  constructor(db: Database, guildId: Model.Snowflake) {
+    if (db == null) throw new Error("db: " + String(db));
+    this._db = db;
+    this._initialize(guildId);
+  }
+
+  _transaction(queries: () => void) {
+    const db = this._db;
+    if (db.inTransaction) {
+      throw new Error("already in transaction");
+    }
+    try {
+      db.prepare("BEGIN").run();
+      queries();
+      if (db.inTransaction) {
+        db.prepare("COMMIT").run();
+      }
+    } finally {
+      if (db.inTransaction) {
+        db.prepare("ROLLBACK").run();
+      }
+    }
+  }
+
+  _initialize(guild: Model.Snowflake) {
+    const db = this._db;
+    // We store the config in a singleton table `meta`, whose unique row
+    // has primary key `0`. Only the first ever insert will succeed; we
+    // are locked into the first config.
+    db.prepare(
+      dedent`\
+        CREATE TABLE IF NOT EXISTS meta (
+            zero INTEGER PRIMARY KEY,
+            config TEXT NOT NULL
+        )
+      `
+    ).run();
+
+    const config = stringify({
+      version: VERSION,
+      guild,
+    });
+
+    const existingConfig: string | void = db
+      .prepare("SELECT config FROM meta")
+      .pluck()
+      .get();
+    if (existingConfig === config) {
+      // Already set up; nothing to do.
+      return;
+    } else if (existingConfig !== undefined) {
+      throw new Error(
+        "Database already populated with incompatible server or version"
+      );
+    }
+    db.prepare("INSERT INTO meta (zero, config) VALUES (0, ?)").run(config);
+
+    const tables = [
+      dedent`\
+        CREATE TABLE channels (
+            id TEXT PRIMARY KEY,
+            type TEXT NOT NULL,
+            name TEXT NOT NULL
+        )
+      `,
+      dedent`\
+        CREATE TABLE members (
+            user_id TEXT PRIMARY KEY,
+            username TEXT NOT NULL,
+            discriminator TEXT NOT NULL,
+            bot INTEGER NOT NULL,
+            nick TEXT
+        )
+      `,
+      dedent`\
+        CREATE TABLE messages (
+            id TEXT PRIMARY KEY,
```

Comment:
(per @Beanow's observation above, we should change all the `TEXT PRIMARY KEY`s to include `NOT NULL`)